contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,308 @@
+        // calculate amount out from swap
+        // @dev reverts for invalid token indices
+        dy_ = calculateSwap(tokenIndexFrom, tokenIndexTo, dx);
+        require(dy_ >= minDy, "dy < minDy");
+
+        // transfer dx in and send dy out
+        address tokenIn = tokenIndexFrom == 0 ? token0 : token1;
+        address tokenOut = tokenIndexTo == 0 ? token0 : token1;
+        IERC20(tokenIn).safeTransferFrom(msg.sender, address(this), dx);
+        IERC20(tokenOut).safeTransfer(msg.sender, dy_);
+
+        emit TokenSwap(msg.sender, dx, dy_, tokenIndexFrom, tokenIndexTo);
+    }
+
+    /// @notice Calculates amount of tokens received on swap
+    /// @dev Reverts if either token index is invalid
+    /// @param tokenIndexFrom The index of the token in
+    /// @param tokenIndexTo The index of the token out
+    /// @param dx The amount of token in in token decimals
+    function calculateSwap(
+        uint8 tokenIndexFrom,
+        uint8 tokenIndexTo,
+        uint256 dx
+    ) public view onlyToken(tokenIndexFrom) onlyToken(tokenIndexTo) returns (uint256 dy_) {
+        require(tokenIndexFrom != tokenIndexTo, "invalid token swap");
+
+        // get current token balances in pool
+        uint256 xWad = _amountWad(IERC20(token0).balanceOf(address(this)), true);
+        uint256 yWad = _amountWad(IERC20(token1).balanceOf(address(this)), false);
+
+        // convert to an amount in wad
+        uint256 amountInWad = _amountWad(dx, tokenIndexFrom == 0);
+
+        // calculate swap amount out wad
+        // @dev obeys P * x + y = D
+        uint256 amountOutWad;
+        if (tokenIndexFrom == 0) {
+            // get D balance before swap
+            uint256 _d = _D(xWad, yWad);
+
+            // token0 in for token1 out
+            xWad += amountInWad;
+
+            // check amount out won't exceed pool balance
+            uint256 prod = Math.mulDiv(P, xWad, wad);
+            require(_d >= prod, "dy > pool balance");
+
+            uint256 yWadAfter = _d - prod;
+            amountOutWad = yWad - yWadAfter;
+        } else {
+            // get D balance before swap
+            uint256 _d = _D(xWad, yWad);
+
+            // token1 in for token0 out
+            yWad += amountInWad;
+
+            // check amount out won't exceed pool balance
+            require(_d >= yWad, "dy > pool balance");
+
+            uint256 xWadAfter = Math.mulDiv(_d - yWad, wad, P);
+            amountOutWad = xWad - xWadAfter;
+        }
+
+        // apply swap fee on amount out
+        amountOutWad -= Math.mulDiv(amountOutWad, fee, wad);
+
+        // convert amount out to decimals
+        dy_ = _amountDecimals(amountOutWad, tokenIndexTo == 0);
+    }
+
+    /// @notice Address of the pooled token at given index
+    /// @dev Reverts for invalid token index
+    /// @param index The index of the token
+    function getToken(uint8 index) external view onlyToken(index) returns (IERC20) {
```

Comment:
`getToken()` reverting when out of bounds is expected behavior btw, so no need to adjust anything here (re: `calculateSwap` comment).